contribute-to-tidb/review-a-pr-stage.md

```diff
@@ -0,0 +1,238 @@
+
+Sometimes, making choices of accepting or rejecting a PR can be tricky as in the following situations: 
+
+* Suppose that a PR contains special optimization that can improve the overall performance by 30%. However, the PR introduces a totally different code path, and every subsequent feature must consider it.
+* Suppose that a PR is to fix a critical bug, but the change in the PR is risky to introduce other bugs.
+
+If a PR under your review is in these tricky situations, what is the right choice, accepting the PR or rejecting it? The answer is always "it depends." Software design is more like a kind of art than technology. It is about aesthetics and your taste of the code. There are always trade-offs, and often there's no perfect solution. Still, there is a general standard (from [@shenli](https://github.com/shenli)). Maybe we call it, "Better than before?":
+
+* The change is making something better, although not a perfect solution.
+* The change can mitigate the pains of users.
+* The change does not introduce a bug.
+
+Then, we can accept the code.
+
+Besides, you should also check some minimal requirements of code reviews.
+
+If a bug fix PR doesn't include a test case, the review should ask the author to add tests.
+
+If a performance improvement PR doesn't include a benchmark result, the reviewer should ask the author to attach one.
+
+## What to check in a code review
+
+### Design
+
+The most important thing of a review is checking the overall design.
+
+To learn the overall design of a PR, you can read the description of the PR. If the description is unclear or hard to understand, you can request the author to improve it. You may consider asking questions such as the following:
+
+* Does the change integrate well with the rest of our system?
+* Is now a good time to add this functionality?
+
+If you need a further discussion of the PR, you can go to [TiDB Internals forum](https://internals.tidb.io/). For pull requests that are created without prior consensus, try to seek consensus there too.
+
+### Functionality
+
+* Does this PR implement what the PR author intends to do?
+* How will implementing this PR help the user?
+
+Basically, PR authors are supposed to sufficiently test their PRs before code reviews. However, as a reviewer, you should still think about edge cases, especially for problems about concurrency.
+
+You can always validate the PR if you want. 
+
+For some changes such as UI, it is hard to understand how the changes will impact users when you are just reading the code. In this case, if it is too inconvenient to try it yourself, you can ask the PR author to give you a demo of the functionality.
+
+### Logic
+
+When concurrent programming is involved, it's particularly important to think about the logic in code review. These sorts of issues are very hard to detect by just running the code and usually need somebody to think carefully.
```

Comment:
```suggestion
When concurrent programming is involved, it is particularly important to review the logic carefully, because the concurrent issues are very hard to detect by just running the code.
```